packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js

```diff
@@ -236,24 +260,86 @@ const writeGranuleAndFilesViaTransaction = async ({
   );
   // Ensure that we get a granule ID for the files even if the
   // upsert query returned an empty result
-  const granuleCumulusId = await getGranuleCumulusIdFromQueryResultOrLookup({
+  return getGranuleCumulusIdFromQueryResultOrLookup({
     trx,
     queryResult: upsertQueryResult,
     granuleRecord,
   });
+};
 
+/**
+ * Generate file records based on workflow status, write files to
+ * the database, and update granule status if file writes fail
+ *
+ * @param {Object} params
+ * @param {Object} params.files - File objects
+ * @param {number} params.granuleCumulusId
+ *   Cumulus ID of the granule for this file
+ * @param {string} params.workflowStatus - Workflow status
+ * @param {Knex} params.knex - Client to interact with Postgres database
+ * @param {Object} params.granulePgModel - Optional Granule model override
+ * @returns {Promise} - Promise resolved once all file writes resolve
+ */
+const _writeGranuleFiles = async ({
+  files,
+  granuleCumulusId,
+  workflowStatus,
+  knex,
+  granulePgModel = new GranulePgModel(),
+}) => {
   let fileRecords = [];
 
   if (workflowStatus !== 'running') {
-    fileRecords = await generateFileRecords({
-      files: updatedFiles,
+    fileRecords = _generateFileRecords({
+      files: files,
       granuleCumulusId,
     });
   }
 
-  return writeFilesViaTransaction({
-    fileRecords,
-    trx,
+  try {
+    await _writeFiles({
+      fileRecords,
+      knex,
+    });
+  } catch (error_) {
+    const granule = await granulePgModel.get(knex, { cumulus_id: granuleCumulusId });
+
+    granulePgModel.upsert(
+      knex,
+      {
+        ...granule,
+        status: 'failed',
+        error: {
+          Error: 'Failed writing files to Postgres.',
+          Cause: error_.message,
+        },
+      }
+    );
+  }
+};
+
+/**
+ * Create Postgres File objects from a granule's files
```

Comment:
The returned files here aren't Postgres file objects. They're just File API objects that have been fleshed out with some extra properties that they may have been missing

```suggestion
 * Transform granule files to latest file API structure
```